File: Leetcode/LinkedList/MergeKSortedLists/MergeKSortedLists.py
# 23. Merge k Sorted Lists
# https://leetcode.com/problems/merge-k-sorted-lists/

from heapq import merge
from typing import List, Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


"""
공간 O(1) 해법
Heap을 쓰는 풀이는 Heap 폴더에 있음.
"""

# O(m*n) 머지 소트 186 ms
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        self.lists = lists
        return self.merge(0, len(lists) - 1)

    def merge(self, left: int, right: int) -> ListNode:
        if left > right:
            return None
        if left == right:
            return self.lists[left]

        mid = (left + right) // 2

        leftNode = self.merge(left, mid)
        rightNode = self.merge(mid + 1, right)

        return self.mergeLists(leftNode, rightNode)

    def mergeLists(self, list1: ListNode, list2: ListNode) -> ListNode:
        head = ListNode()
        cur = head

        while list1 and list2:
            if list1.val <= list2.val:
                cur.next = list1
                list1 = list1.next
            else:
                cur.next = list2
                list2 = list2.next
            cur = cur.next

        if list1:
            cur.next = list1
        elif list2:
            cur.next = list2

        return head.next


# O(m*n) 하나씩 합치기 반복 5359 ms
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        head = ListNode()
        cur = head

        for list in lists:
            cur.next = self.mergeLists(cur.next, list)

        return head.next

    def mergeLists(self, list1, list2):
        head = ListNode()
        cur = head

        while list1 and list2:
            if list1.val <= list2.val:
                cur.next = list1
                list1 = list1.next
            else:
                cur.next = list2
                list2 = list2.next
            cur = cur.next

        if list1:
            cur.next = list1
        if list2:
            cur.next = list2

        return head.next


# O(m*n), m : length of Lists, n : length of linked-list
# for문으로 한번에 합치기 6828 ms
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        head = ListNode()
        cur = head

        while any(lists):
            minVal = 10001
            num = -1
            for i, list in enumerate(lists):
                if not list:
                    continue
                if minVal > list.val:
                    minVal = list.val
                    num = i

            cur.next = lists[num]
            cur = cur.next
            lists[num] = lists[num].next

        return head.next

File: Leetcode/LinkedList/MergeKSortedLists/MergeKSortedLists.java
package LinkedList.MergeKSortedLists;
// 23. Merge k Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists/

import java.util.Arrays;
// Definition for singly-linked list.
import LinkedList.ListNode;


// 분할 머지 빠름 3ms.
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        return merge(lists, 0, lists.length - 1);
    }

    private ListNode merge(ListNode[] lists, int left, int right) {
        if (left > right) {
            return null;
        }
        if (left == right) {
            return lists[left];
        }

        int mid = (left + right) / 2;

        ListNode leftNode = merge(lists, left, mid);
        ListNode rightNode = merge(lists, mid + 1, right);

        return mergeLists(leftNode, rightNode);
    }

    private ListNode mergeLists(ListNode node1, ListNode node2) {
        ListNode head = new ListNode();
        ListNode dummy = head;

        while (node1 != null && node2 != null) {
            if (node1.val <= node2.val) {
                dummy.next = node1;
                node1 = node1.next;
            } else {
                dummy.next = node2;
                node2 = node2.next;
            }
            dummy = dummy.next;
        }

        if (node1 != null) {
            dummy.next = node1;
        } else if (node2 != null) {
            dummy.next = node2;
        }

        return head.next;
    }
}


// 단순하게 앞에서부터 머지 느림 248ms.
class Solution1 {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0)
            return null;

        ListNode dummy = null;
        for (ListNode node : lists) {
            dummy = mergeLists(dummy, node);
        }

        return dummy;
    }

    private ListNode mergeLists(ListNode node1, ListNode node2) {
        ListNode head = new ListNode();
        ListNode dummy = head;

        while (node1 != null && node2 != null) {
            if (node1.val <= node2.val) {
                dummy.next = node1;
                node1 = node1.next;
            } else {
                dummy.next = node2;
                node2 = node2.next;
            }
            dummy = dummy.next;
        }

        if (node1 != null) {
            dummy.next = node1;
        } else if (node2 != null) {
            dummy.next = node2;
        }

        return head.next;
    }
}


public class MergeKSortedLists {
    public static void main(String[] args) {
        testSol(new ListNode[] {new ListNode(new int[] {1, 4, 5}),
                new ListNode(new int[] {1, 3, 4}), new ListNode(new int[] {2, 6})},
                new int[] {1, 1, 2, 3, 4, 4, 5, 6});
        testSol(new ListNode[] {}, null);
    }

    static void testSol(ListNode[] input, int[] output) {
        Solution sol = new Solution();
        // todo : solution match
        ListNode temp = sol.mergeKLists(input);
        int[] res = temp != null ? temp.toIntArrays() : null;
        if (Arrays.equals(res, output)) {
            System.out.println("O : " + res);
        } else {
            System.out.println("x : " + res + "	expect : " + output);
        }
    }
}
